{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split, StratifiedKFold\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import f1_score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Support Vector Classification\n",
    "\n",
    "In this notebook, we explore using support vector classifiers for retracted and non-retracted papers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>publication_year</th>\n",
       "      <th>countries_distinct_count</th>\n",
       "      <th>institutions_distinct_count</th>\n",
       "      <th>referenced_works_count</th>\n",
       "      <th>cited_by_count</th>\n",
       "      <th>authors_distinct_count</th>\n",
       "      <th>any_author_has_retraction</th>\n",
       "      <th>min_retracted_author_rank</th>\n",
       "      <th>...</th>\n",
       "      <th>has_10pct_retracted_author</th>\n",
       "      <th>top_percentile_retracted_author</th>\n",
       "      <th>frac_author_repeat_offenders</th>\n",
       "      <th>any_institution_has_retraction</th>\n",
       "      <th>min_retracted_institution_rank</th>\n",
       "      <th>has_1pct_retracted_institution</th>\n",
       "      <th>has_5pct_retracted_institution</th>\n",
       "      <th>has_10pct_retracted_institution</th>\n",
       "      <th>top_percentile_retracted_institution</th>\n",
       "      <th>is_retracted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://openalex.org/W2031611770</td>\n",
       "      <td>FastTree 2 – Approximately Maximum-Likelihood ...</td>\n",
       "      <td>2010</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>10411</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>-inf</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>284.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>95.046385</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://openalex.org/W2041257508</td>\n",
       "      <td>progressiveMauve: Multiple Genome Alignment wi...</td>\n",
       "      <td>2010</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>62</td>\n",
       "      <td>3254</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>-inf</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>257.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>95.518992</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://openalex.org/W2169773990</td>\n",
       "      <td>Rapid SNP Discovery and Genetic Mapping Using ...</td>\n",
       "      <td>2008</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>2920</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>-inf</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>843.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>85.261684</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://openalex.org/W2071754162</td>\n",
       "      <td>Source Partitioning Using Stable Isotopes: Cop...</td>\n",
       "      <td>2010</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>2321</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>-inf</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>414.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>92.762121</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://openalex.org/W2135989088</td>\n",
       "      <td>Gut Microbiota in Human Adults with Type 2 Dia...</td>\n",
       "      <td>2010</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>2293</td>\n",
       "      <td>10</td>\n",
       "      <td>True</td>\n",
       "      <td>4773.5</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>88.490293</td>\n",
       "      <td>0.1</td>\n",
       "      <td>True</td>\n",
       "      <td>233.5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>95.930334</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 id  \\\n",
       "0  https://openalex.org/W2031611770   \n",
       "1  https://openalex.org/W2041257508   \n",
       "2  https://openalex.org/W2169773990   \n",
       "3  https://openalex.org/W2071754162   \n",
       "4  https://openalex.org/W2135989088   \n",
       "\n",
       "                                               title publication_year  \\\n",
       "0  FastTree 2 – Approximately Maximum-Likelihood ...             2010   \n",
       "1  progressiveMauve: Multiple Genome Alignment wi...             2010   \n",
       "2  Rapid SNP Discovery and Genetic Mapping Using ...             2008   \n",
       "3  Source Partitioning Using Stable Isotopes: Cop...             2010   \n",
       "4  Gut Microbiota in Human Adults with Type 2 Dia...             2010   \n",
       "\n",
       "  countries_distinct_count institutions_distinct_count referenced_works_count  \\\n",
       "0                        1                           2                     30   \n",
       "1                        1                           2                     62   \n",
       "2                        1                           1                     18   \n",
       "3                        2                           3                     24   \n",
       "4                        1                           2                     36   \n",
       "\n",
       "  cited_by_count authors_distinct_count  any_author_has_retraction  \\\n",
       "0          10411                      3                      False   \n",
       "1           3254                      3                      False   \n",
       "2           2920                      9                      False   \n",
       "3           2321                      4                      False   \n",
       "4           2293                     10                       True   \n",
       "\n",
       "   min_retracted_author_rank  ...  has_10pct_retracted_author  \\\n",
       "0                       -inf  ...                       False   \n",
       "1                       -inf  ...                       False   \n",
       "2                       -inf  ...                       False   \n",
       "3                       -inf  ...                       False   \n",
       "4                     4773.5  ...                        True   \n",
       "\n",
       "   top_percentile_retracted_author  frac_author_repeat_offenders  \\\n",
       "0                         0.000000                           0.0   \n",
       "1                         0.000000                           0.0   \n",
       "2                         0.000000                           0.0   \n",
       "3                         0.000000                           0.0   \n",
       "4                        88.490293                           0.1   \n",
       "\n",
       "   any_institution_has_retraction  min_retracted_institution_rank  \\\n",
       "0                            True                           284.0   \n",
       "1                            True                           257.0   \n",
       "2                            True                           843.0   \n",
       "3                            True                           414.5   \n",
       "4                            True                           233.5   \n",
       "\n",
       "   has_1pct_retracted_institution  has_5pct_retracted_institution  \\\n",
       "0                           False                            True   \n",
       "1                           False                            True   \n",
       "2                           False                           False   \n",
       "3                           False                           False   \n",
       "4                           False                            True   \n",
       "\n",
       "   has_10pct_retracted_institution  top_percentile_retracted_institution  \\\n",
       "0                             True                             95.046385   \n",
       "1                             True                             95.518992   \n",
       "2                            False                             85.261684   \n",
       "3                             True                             92.762121   \n",
       "4                             True                             95.930334   \n",
       "\n",
       "   is_retracted  \n",
       "0         False  \n",
       "1         False  \n",
       "2         False  \n",
       "3         False  \n",
       "4         False  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data  = pd.read_pickle('./Data/OpenAlex/openalex-data-plosone-2000-2010-train-04.pkl')\n",
    "data.dropna\n",
    "#data=data.drop('https://openalex.org/W2031611770')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "reduced_data = data.drop(columns=['id', 'title', 'min_retracted_author_rank', 'min_retracted_institution_rank'])\n",
    "\n",
    "for col in reduced_data.columns:\n",
    "    if reduced_data[col].dtype == bool:\n",
    "        reduced_data[col] = reduced_data[col].astype(int)\n",
    "    elif col in ['publication_year', 'countries_distinct_count', 'institutions_distinct_count', 'referenced_works_count', 'cited_by_count', 'authors_distinct_count']:\n",
    "        reduced_data[col] = reduced_data[col].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test = train_test_split(reduced_data.copy(), test_size = 0.2, random_state=123, stratify=reduced_data['is_retracted'])\n",
    "\n",
    "y_train, y_test = X_train['is_retracted'].values, X_test['is_retracted'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in X_train.columns:\n",
    "    if X_train[col].dtype == float:\n",
    "        mean = X_train[col].mean()\n",
    "        std = X_train[col].std()\n",
    "        X_train[col] = (X_train[col] - mean)/std\n",
    "        X_test[col] = (X_test[col] - mean)/std  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model Selection\n",
    "\n",
    "Using k-fold cross validation to determine the best kernel for our support vector machine. We will test linear, polynomial, rbf, and sigmoid kernels. For polynomial kernels, we test degrees 2, 3, 5, 10."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2856     0\n",
       "5098     0\n",
       "14499    0\n",
       "6475     0\n",
       "11873    0\n",
       "        ..\n",
       "12410    0\n",
       "11151    0\n",
       "1948     0\n",
       "2110     0\n",
       "15368    0\n",
       "Name: is_retracted, Length: 12647, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train['is_retracted']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "kfold = StratifiedKFold(n_splits = 10, shuffle=True, random_state=123)\n",
    "\n",
    "f1_scores = np.zeros((14, 10))\n",
    "\n",
    "features = list(X_train.columns)\n",
    "features.remove('is_retracted')\n",
    "\n",
    "i = 0\n",
    "\n",
    "for train_index, test_index in kfold.split(X=X_train[features], y=X_train['is_retracted']):\n",
    "    X_tt = X_train[features].iloc[train_index]\n",
    "    y_tt = X_train['is_retracted'].iloc[train_index].values\n",
    "    X_val = X_train[features].iloc[test_index]\n",
    "    y_val = X_train['is_retracted'].iloc[test_index].values\n",
    "    \n",
    "\n",
    "    model = SVC(kernel='linear', gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[0,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 2, gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[1,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 3, gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[2,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 5, gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[3,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 10, gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[4,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='rbf', gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[5,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='sigmoid', gamma='auto')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[6,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='linear')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[7,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 2)\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[8,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 3)\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[9,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 5)\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[10,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='poly', degree = 10)\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[11,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='rbf')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[12,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    model = SVC(kernel='sigmoid')\n",
    "    model.fit(X_tt, y_tt)\n",
    "    f1_scores[13,i] = f1_score(y_val, model.predict(X_val))\n",
    "\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kernel 0 Accuracy: 0.40952380952380957\n",
      "Kernel 1 Accuracy: 0.45357142857142857\n",
      "Kernel 2 Accuracy: 0.5754761904761905\n",
      "Kernel 3 Accuracy: 0.505\n",
      "Kernel 4 Accuracy: 0.43753968253968256\n",
      "Kernel 5 Accuracy: 0.43380952380952376\n",
      "Kernel 6 Accuracy: 0.0\n",
      "Kernel 7 Accuracy: 0.40952380952380957\n",
      "Kernel 8 Accuracy: 0.5083333333333333\n",
      "Kernel 9 Accuracy: 0.5254761904761905\n",
      "Kernel 10 Accuracy: 0.4719913419913421\n",
      "Kernel 11 Accuracy: 0.46992063492063485\n",
      "Kernel 12 Accuracy: 0.39380952380952383\n",
      "Kernel 13 Accuracy: 0.0\n"
     ]
    }
   ],
   "source": [
    "for i in range(14):\n",
    "    print('Kernel', i, 'Accuracy:',np.mean(f1_scores[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Kernel 2, the polynomial kernel with degree 3 and gamma = auto, is giving us the best F1 score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
